module/apmgoredis/client.go

```diff
@@ -0,0 +1,163 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package apmgoredis
+
+import (
+	"context"
+	"strings"
+
+	"github.com/go-redis/redis"
+
+	"go.elastic.co/apm"
+)
+
+// Client is the interface returned by Wrap.
+//
+// Client implements redis.UniversalClient
+type Client interface {
+	redis.UniversalClient
+
+	// ClusterClient returns the wrapped *redis.ClusterClient,
+	// or nil if a non-cluster client is wrapped.
+	Cluster() *redis.ClusterClient
+
+	// Ring returns the wrapped *redis.Ring,
+	// or nil if a non-ring client is wrapped.
+	RingClient() *redis.Ring
+
+	// WithContext returns a shallow copy of the client with
+	// its context changed to ctx and will add instrumentation
+	// with client.WrapProcess and client.WrapProcessPipeline
+	//
+	// To report commands as spans, ctx must contain a transaction or span.
+	WithContext(ctx context.Context) Client
+}
+
+// Wrap wraps client such that executed commands are reported as spans to Elastic APM,
+// using the client's associated context.
+// A context-specific client may be obtained by using Client.WithContext.
```

Comment:
It appears from the tests that not everything works with Ring. Can you please add a comment here explaining what isn't supported, linking to issues if there are any?

Reply:
`Ring` has not `TxPipeline/TxPipelined` implemented, it's not related to the instrumentation with the apm agent, do you want me to add a comment for this?

While, related to instrumentation, it works with `Ring` on `Pipeline/Pipelined` but not for direct command on the client: this PR https://github.com/go-redis/redis/pull/1017 will address

For me it's ok to wait for that PR to be merged, and then remove the special case in the test here

Reply:
I see, thanks. I'd prefer to wait for go-redis/redis#1017 to merge first, to avoid a little bit of churn.